File: mRemoteNG/Forms/OptionsPages/KeyboardPage.vb
Imports mRemoteNG.Config
Imports mRemoteNG.My

Namespace Forms.OptionsPages
    Public Class KeyboardPage
        Public Overrides Property PageName() As String
            Get
                Return Language.strOptionsTabKeyboard
            End Get
            Set(value As String)
            End Set
        End Property

        Public Overrides Sub ApplyLanguage()
            MyBase.ApplyLanguage()

            lblKeyboardShortcuts.Text = Language.strOptionsKeyboardLabelKeyboardShortcuts
            btnNewKeyboardShortcut.Text = Language.strOptionsKeyboardButtonNew
            btnDeleteKeyboardShortcut.Text = Language.strOptionsKeyboardButtonDelete
            btnResetKeyboardShortcuts.Text = Language.strOptionsKeyboardButtonReset
            grpModifyKeyboardShortcut.Text = Language.strOptionsKeyboardGroupModifyShortcut
            btnResetAllKeyboardShortcuts.Text = Language.strOptionsKeyboardButtonResetAll
        End Sub

        Public Overrides Sub LoadSettings()
            _tabsListViewGroup = New ListViewGroup(Language.strOptionsKeyboardCommandsGroupTabs)
            _previousTabListViewItem = New ListViewItem(Language.strOptionsKeyboardCommandsPreviousTab, _tabsListViewGroup)
            _nextTabListViewItem = New ListViewItem(Language.strOptionsKeyboardCommandsNextTab, _tabsListViewGroup)

            _keyboardShortcutMap = KeyboardShortcuts.Map.Clone()

            lvKeyboardCommands.Groups.Add(_tabsListViewGroup)
            lvKeyboardCommands.Items.Add(_previousTabListViewItem)
            lvKeyboardCommands.Items.Add(_nextTabListViewItem)
            _previousTabListViewItem.Selected = True
        End Sub

        Public Overrides Sub SaveSettings()
            MyBase.SaveSettings()

            If _keyboardShortcutMap IsNot Nothing Then
                KeyboardShortcuts.Map = _keyboardShortcutMap
            End If
        End Sub

#Region "Private Fields"
        Private _keyboardShortcutMap As KeyboardShortcutMap
        Private _tabsListViewGroup As ListViewGroup
        Private _previousTabListViewItem As ListViewItem
        Private _nextTabListViewItem As ListViewItem
        Private _ignoreKeyboardShortcutTextChanged As Boolean = False
#End Region

#Region "Private Methods"
#Region "Event Handlers"
        Private Sub lvKeyboardCommands_SelectedIndexChanged(sender As System.Object, e As EventArgs) Handles lvKeyboardCommands.SelectedIndexChanged
            Dim isItemSelected As Boolean = (lvKeyboardCommands.SelectedItems.Count = 1)
            EnableKeyboardShortcutControls(isItemSelected)

            If Not isItemSelected Then Return

            Dim selectedItem As ListViewItem = lvKeyboardCommands.SelectedItems(0)

            lblKeyboardCommand.Text = selectedItem.Text
            lstKeyboardShortcuts.Items.Clear()

            lstKeyboardShortcuts.Items.AddRange(_keyboardShortcutMap.GetShortcutKeys(GetSelectedShortcutCommand()))

            If lstKeyboardShortcuts.Items.Count > 0 Then lstKeyboardShortcuts.SelectedIndex = 0
        End Sub

        Private Sub lstKeyboardShortcuts_SelectedIndexChanged(sender As System.Object, e As EventArgs) Handles lstKeyboardShortcuts.SelectedIndexChanged
            Dim isItemSelected As Boolean = (lstKeyboardShortcuts.SelectedItems.Count = 1)

            btnDeleteKeyboardShortcut.Enabled = isItemSelected
            grpModifyKeyboardShortcut.Enabled = isItemSelected
            hotModifyKeyboardShortcut.Enabled = isItemSelected

            If Not isItemSelected Then
                hotModifyKeyboardShortcut.Text = String.Empty
                Return
            End If

            Dim selectedItem As Object = lstKeyboardShortcuts.SelectedItems(0)
            Dim shortcutKey As ShortcutKey = TryCast(selectedItem, ShortcutKey)
            If shortcutKey Is Nothing Then Return

            Dim keysValue As Keys = shortcutKey
            Dim keyCode As Keys = keysValue And Keys.KeyCode
            Dim modifiers As Keys = keysValue And Keys.Modifiers

            _ignoreKeyboardShortcutTextChanged = True
            hotModifyKeyboardShortcut.KeyCode = keyCode
            hotModifyKeyboardShortcut.HotkeyModifiers = modifiers
            _ignoreKeyboardShortcutTextChanged = False
        End Sub

        Private Sub btnNewKeyboardShortcut_Click(sender As System.Object, e As EventArgs) Handles btnNewKeyboardShortcut.Click
            For Each item As Object In lstKeyboardShortcuts.Items
                Dim shortcutKey As ShortcutKey = TryCast(item, ShortcutKey)
                If shortcutKey Is Nothing Then Continue For
                If shortcutKey = 0 Then
                    lstKeyboardShortcuts.SelectedItem = item
                    Return
                End If
            Next

            lstKeyboardShortcuts.SelectedIndex = lstKeyboardShortcuts.Items.Add(New ShortcutKey(Keys.None))
            hotModifyKeyboardShortcut.Focus()
        End Sub

        Private Sub btnDeleteKeyboardShortcut_Click(sender As System.Object, e As EventArgs) Handles btnDeleteKeyboardShortcut.Click
            Dim selectedIndex As Integer = lstKeyboardShortcuts.SelectedIndex

            Dim command As ShortcutCommand = GetSelectedShortcutCommand()
            Dim key As ShortcutKey = TryCast(lstKeyboardShortcuts.SelectedItem, ShortcutKey)
            If Not command = ShortcutCommand.None And key IsNot Nothing Then
                _keyboardShortcutMap.Remove(GetSelectedShortcutCommand(), key)
            End If

            lstKeyboardShortcuts.Items.Remove(lstKeyboardShortcuts.SelectedItem)

            If selectedIndex >= lstKeyboardShortcuts.Items.Count Then selectedIndex = lstKeyboardShortcuts.Items.Count - 1
            lstKeyboardShortcuts.SelectedIndex = selectedIndex
        End Sub

        Private Sub btnResetAllKeyboardShortcuts_Click(sender As System.Object, e As EventArgs) Handles btnResetAllKeyboardShortcuts.Click
            _keyboardShortcutMap = KeyboardShortcuts.DefaultMap.Clone()
            lvKeyboardCommands_SelectedIndexChanged(Me, New EventArgs())
        End Sub

        Private Sub btnResetKeyboardShortcuts_Click(sender As System.Object, e As EventArgs) Handles btnResetKeyboardShortcuts.Click
            Dim command As ShortcutCommand = GetSelectedShortcutCommand()
            If command = ShortcutCommand.None Then Return
            _keyboardShortcutMap.SetShortcutKeys(command, KeyboardShortcuts.DefaultMap.GetShortcutKeys(command))
            lvKeyboardCommands_SelectedIndexChanged(Me, New EventArgs())
        End Sub

        Private Sub hotModifyKeyboardShortcut_TextChanged(sender As System.Object, e As EventArgs) Handles hotModifyKeyboardShortcut.TextChanged
            If _ignoreKeyboardShortcutTextChanged Or _
                lstKeyboardShortcuts.SelectedIndex < 0 Or _
                lstKeyboardShortcuts.SelectedIndex >= lstKeyboardShortcuts.Items.Count Then Return

            Dim keysValue As Keys = (hotModifyKeyboardShortcut.KeyCode And Keys.KeyCode) Or _
                                    (hotModifyKeyboardShortcut.HotkeyModifiers And Keys.Modifiers)

            Dim hadFocus As Boolean = hotModifyKeyboardShortcut.ContainsFocus

            Dim command As ShortcutCommand = GetSelectedShortcutCommand()
            Dim newShortcutKey As New ShortcutKey(keysValue)

            If Not command = ShortcutCommand.None Then
                Dim oldShortcutKey As ShortcutKey = TryCast(lstKeyboardShortcuts.SelectedItem, ShortcutKey)
                If oldShortcutKey IsNot Nothing Then
                    _keyboardShortcutMap.Remove(command, oldShortcutKey)
                End If
                _keyboardShortcutMap.Add(command, newShortcutKey)
            End If

            lstKeyboardShortcuts.Items(lstKeyboardShortcuts.SelectedIndex) = newShortcutKey

            If hadFocus Then
                hotModifyKeyboardShortcut.Focus()
                hotModifyKeyboardShortcut.Select(hotModifyKeyboardShortcut.TextLength, 0)
            End If
        End Sub
#End Region

        Private Function GetSelectedShortcutCommand() As ShortcutCommand
            If Not (lvKeyboardCommands.SelectedItems.Count = 1) Then Return ShortcutCommand.None

            Dim selectedItem As ListViewItem = lvKeyboardCommands.SelectedItems(0)
            If selectedItem Is _previousTabListViewItem Then
                Return ShortcutCommand.PreviousTab
            ElseIf selectedItem Is _nextTabListViewItem Then
                Return ShortcutCommand.NextTab
            End If
        End Function

        Private Sub EnableKeyboardShortcutControls(Optional ByVal enable As Boolean = True)
            lblKeyboardCommand.Visible = enable
            lblKeyboardShortcuts.Enabled = enable
            lstKeyboardShortcuts.Enabled = enable
            btnNewKeyboardShortcut.Enabled = enable
            btnResetKeyboardShortcuts.Enabled = enable

            If Not enable Then
                btnDeleteKeyboardShortcut.Enabled = False
                grpModifyKeyboardShortcut.Enabled = False
                hotModifyKeyboardShortcut.Enabled = False

                lstKeyboardShortcuts.Items.Clear()
                hotModifyKeyboardShortcut.Text = String.Empty
            End If
        End Sub
#End Region
    End Class
End Namespace

File: mRemoteV1/UI/Forms/OptionsPages/KeyboardPage.cs
using System;
using System.Windows.Forms;
using mRemoteNG.Config.KeyboardShortcuts;
using mRemoteNG.My;

namespace mRemoteNG.UI.Forms.OptionsPages
{
    public partial class KeyboardPage
    {
        private bool _ignoreKeyboardShortcutTextChanged;
        private KeyboardShortcutMap _keyboardShortcutMap;
        private ListViewItem _nextTabListViewItem;
        private ListViewItem _previousTabListViewItem;
        private ListViewGroup _tabsListViewGroup;


        public KeyboardPage()
        {
            InitializeComponent();
        }

        public override string PageName
        {
            get { return Language.strOptionsTabKeyboard; }
            set { }
        }

        public override void ApplyLanguage()
        {
            base.ApplyLanguage();

            lblKeyboardShortcuts.Text = Language.strOptionsKeyboardLabelKeyboardShortcuts;
            btnNewKeyboardShortcut.Text = Language.strOptionsKeyboardButtonNew;
            btnDeleteKeyboardShortcut.Text = Language.strOptionsKeyboardButtonDelete;
            btnResetKeyboardShortcuts.Text = Language.strOptionsKeyboardButtonReset;
            grpModifyKeyboardShortcut.Text = Language.strOptionsKeyboardGroupModifyShortcut;
            btnResetAllKeyboardShortcuts.Text = Language.strOptionsKeyboardButtonResetAll;
        }

        public override void LoadSettings()
        {
            _tabsListViewGroup = new ListViewGroup(Language.strOptionsKeyboardCommandsGroupTabs);
            _previousTabListViewItem = new ListViewItem(Language.strOptionsKeyboardCommandsPreviousTab,
                _tabsListViewGroup);
            _nextTabListViewItem = new ListViewItem(Language.strOptionsKeyboardCommandsNextTab, _tabsListViewGroup);

            _keyboardShortcutMap = (KeyboardShortcutMap) KeyboardShortcuts.Map.Clone();

            lvKeyboardCommands.Groups.Add(_tabsListViewGroup);
            lvKeyboardCommands.Items.Add(_previousTabListViewItem);
            lvKeyboardCommands.Items.Add(_nextTabListViewItem);
            _previousTabListViewItem.Selected = true;
        }

        public override void SaveSettings()
        {
            base.SaveSettings();

            if (_keyboardShortcutMap != null)
            {
                KeyboardShortcuts.Map = _keyboardShortcutMap;
            }
        }

        #region Private Methods

        #region Event Handlers

        public void lvKeyboardCommands_SelectedIndexChanged(object sender, EventArgs e)
        {
            var isItemSelected = lvKeyboardCommands.SelectedItems.Count == 1;
            EnableKeyboardShortcutControls(isItemSelected);

            if (!isItemSelected)
            {
                return;
            }

            var selectedItem = lvKeyboardCommands.SelectedItems[0];

            lblKeyboardCommand.Text = selectedItem.Text;
            lstKeyboardShortcuts.Items.Clear();

            lstKeyboardShortcuts.Items.AddRange(_keyboardShortcutMap.GetShortcutKeys(GetSelectedShortcutCommand()));

            if (lstKeyboardShortcuts.Items.Count > 0)
            {
                lstKeyboardShortcuts.SelectedIndex = 0;
            }
        }

        public void lstKeyboardShortcuts_SelectedIndexChanged(object sender, EventArgs e)
        {
            var isItemSelected = lstKeyboardShortcuts.SelectedItems.Count == 1;

            btnDeleteKeyboardShortcut.Enabled = isItemSelected;
            grpModifyKeyboardShortcut.Enabled = isItemSelected;
            hotModifyKeyboardShortcut.Enabled = isItemSelected;

            if (!isItemSelected)
            {
                hotModifyKeyboardShortcut.Text = string.Empty;
                return;
            }

            var selectedItem = lstKeyboardShortcuts.SelectedItems[0];
            var shortcutKey = selectedItem as ShortcutKey;
            if (shortcutKey == null)
            {
                return;
            }

            //Keys keysValue = System.Windows.Forms.Keys.A;
            var keyCode = Keys.A;
            var modifiers = Keys.A;

            _ignoreKeyboardShortcutTextChanged = true;
            hotModifyKeyboardShortcut.KeyCode = keyCode;
            hotModifyKeyboardShortcut.HotkeyModifiers = modifiers;
            _ignoreKeyboardShortcutTextChanged = false;
        }

        public void btnNewKeyboardShortcut_Click(object sender, EventArgs e)
        {
            foreach (var item in lstKeyboardShortcuts.Items)
            {
                var shortcutKey = item as ShortcutKey;
                if (shortcutKey?.KeyCode == 0)
                {
                    lstKeyboardShortcuts.SelectedItem = item;
                    return;
                }
            }

            lstKeyboardShortcuts.SelectedIndex = lstKeyboardShortcuts.Items.Add(new ShortcutKey(Keys.None));
            hotModifyKeyboardShortcut.Focus();
        }

        public void btnDeleteKeyboardShortcut_Click(object sender, EventArgs e)
        {
            var selectedIndex = lstKeyboardShortcuts.SelectedIndex;

            var command = GetSelectedShortcutCommand();
            var key = lstKeyboardShortcuts.SelectedItem as ShortcutKey;
            if (command != ShortcutCommand.None & key != null)
            {
                _keyboardShortcutMap.Remove(GetSelectedShortcutCommand(), key);
            }

            lstKeyboardShortcuts.Items.Remove(lstKeyboardShortcuts.SelectedItem);

            if (selectedIndex >= lstKeyboardShortcuts.Items.Count)
            {
                selectedIndex = lstKeyboardShortcuts.Items.Count - 1;
            }
            lstKeyboardShortcuts.SelectedIndex = selectedIndex;
        }

        public void btnResetAllKeyboardShortcuts_Click(object sender, EventArgs e)
        {
            _keyboardShortcutMap = (KeyboardShortcutMap) KeyboardShortcuts.DefaultMap.Clone();
            lvKeyboardCommands_SelectedIndexChanged(this, new EventArgs());
        }

        public void btnResetKeyboardShortcuts_Click(object sender, EventArgs e)
        {
            var command = GetSelectedShortcutCommand();
            if (command == ShortcutCommand.None)
            {
                return;
            }
            _keyboardShortcutMap.SetShortcutKeys(command, KeyboardShortcuts.DefaultMap.GetShortcutKeys(command));
            lvKeyboardCommands_SelectedIndexChanged(this, new EventArgs());
        }

        public void hotModifyKeyboardShortcut_TextChanged(object sender, EventArgs e)
        {
            if (_ignoreKeyboardShortcutTextChanged ||
                lstKeyboardShortcuts.SelectedIndex < 0 |
                lstKeyboardShortcuts.SelectedIndex >= lstKeyboardShortcuts.Items.Count)
            {
                return;
            }

            var keysValue = Keys.A;

            var hadFocus = hotModifyKeyboardShortcut.ContainsFocus;

            var command = GetSelectedShortcutCommand();
            var newShortcutKey = new ShortcutKey(keysValue);

            if (!(command == ShortcutCommand.None))
            {
                var oldShortcutKey = lstKeyboardShortcuts.SelectedItem as ShortcutKey;
                if (oldShortcutKey != null)
                {
                    _keyboardShortcutMap.Remove(command, oldShortcutKey);
                }
                _keyboardShortcutMap.Add(command, newShortcutKey);
            }

            lstKeyboardShortcuts.Items[lstKeyboardShortcuts.SelectedIndex] = newShortcutKey;

            if (hadFocus)
            {
                hotModifyKeyboardShortcut.Focus();
                hotModifyKeyboardShortcut.Select(hotModifyKeyboardShortcut.TextLength, 0);
            }
        }

        #endregion

        private ShortcutCommand GetSelectedShortcutCommand()
        {
            if (!(lvKeyboardCommands.SelectedItems.Count == 1))
            {
                return ShortcutCommand.None;
            }

            var selectedItem = lvKeyboardCommands.SelectedItems[0];
            if (selectedItem == _previousTabListViewItem)
            {
                return ShortcutCommand.PreviousTab;
            }
            if (selectedItem == _nextTabListViewItem)
            {
                return ShortcutCommand.NextTab;
            }
            return ShortcutCommand.None;
        }

        private void EnableKeyboardShortcutControls(bool enable = true)
        {
            lblKeyboardCommand.Visible = enable;
            lblKeyboardShortcuts.Enabled = enable;
            lstKeyboardShortcuts.Enabled = enable;
            btnNewKeyboardShortcut.Enabled = enable;
            btnResetKeyboardShortcuts.Enabled = enable;
            if (!enable)
            {
                btnDeleteKeyboardShortcut.Enabled = false;
                grpModifyKeyboardShortcut.Enabled = false;
                hotModifyKeyboardShortcut.Enabled = false;
                lstKeyboardShortcuts.Items.Clear();
                hotModifyKeyboardShortcut.Text = string.Empty;
            }
        }

        #endregion
    }
}